src/ballet/bmtree/fd_bmtree.h

```diff
@@ -0,0 +1,284 @@
+#ifndef HEADER_fd_src_ballet_bmtree_fd_bmtree_h
+#define HEADER_fd_src_ballet_bmtree_fd_bmtree_h
+
+#include "../sha256/fd_sha256.h"
+
+/* fd_bmtree{20,32} is a binary Merkle tree used in the Solana protocol.
+   It uses the SHA-256 hash function.
+
+   Specification:
+   https://github.com/solana-foundation/specs/blob/main/core/merkle-tree.md
+
+   It is generally used as a vector commitment scheme
+   wherein the root node of the tree commits the vector of leaf nodes.
+
+   All methods provided by this Merkle tree derive from these three
+   basic operations.
+
+     1. Construct leaf node:
+
+        (leaf blob) -> (node)
+
+     2. Construct branch node with two children:
+
+        (node, node) -> (node)
+
+     3. Construct branch node with one child:
+
+        (node) -> (node)
+
+   Example derived methods.
+   (TODO not all of them are provided by this header file yet)
+
+     4. Construct full tree:
+
+        (vector of leaf blobs) -> (tree of nodes)
+
+     5. Create inclusion proof from tree data
+
+        (tree of nodes, node index) -> (inclusion proof)
+
+     6. Verify node inclusion proof
+
+        (node, root node, inclusion proof) -> (bool)
+
+   **Topology**
+
+   Tree topology has the following constraints:
+
+    - All leaf nodes are in the bottom level
+
+    - If a given layer `L`
+      with number of nodes `L_n` ...
+
+      ... has exactly one node,
+          this one node contains the root node
+          and is the uppermost layer.
+
+      ... has more than one node
+          ... and `L_n % 2 == 0`,
+              the layer above contains n/2 nodes
+
+          ... and `L_n % 2 == 1`,
+              the layer above contains (n/2)+1 nodes.
+
```

Comment:
Can you add something like this:

You can imagine the process for generating the tree with n leaf nodes being the following:
Start with the smallest complete binary tree that has at least n leaf nodes. Label the leaf nodes from left to right L0, L1, ... L(n-1). Delete any un-labeled leaf nodes, and then recursively delete any nodes with no children. For any nodes with a single (un-deleted) child, duplicate the link to the child.
Each non-leaf node now has exactly two children, counting duplicates.